_includes/tutorials/aggregating-minmax/flinksql/code/src/test/java/io/confluent/developer/AbstractFlinkKafkaTest.java

```diff
@@ -0,0 +1,171 @@
+import org.apache.flink.table.api.bridge.java.StreamTableEnvironment;
+import org.assertj.core.util.Sets;
+import org.junit.Before;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.KafkaContainer;
+import org.testcontainers.containers.Network;
+import org.testcontainers.utility.DockerImageName;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.PrintStream;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.Set;
+
+import static org.testcontainers.containers.KafkaContainer.KAFKA_PORT;
+
+/**
+ * Base class for Flink SQL integration tests that use Flink's Kafka connectors. Encapsulates
+ * Kafka broker and Schema Registry Testcontainer management and includes utility methods for
+ * dynamically configuring Flink SQL Kafka connectors and processing Table API results.
+ */
+public class AbstractFlinkKafkaTest {
+
+  protected StreamTableEnvironment tEnv;
+  protected Integer schemaRegistryPort, kafkaPort;
+
+  @Before
+  public void setUp() {
+    // create Flink table environment that test subclasses will use to execute SQL statements
+    StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+    env.setParallelism(4);
+    env.getConfig().setRestartStrategy(RestartStrategies.noRestart());
+    env.setStateBackend(new EmbeddedRocksDBStateBackend());
+    tEnv = StreamTableEnvironment.create(env, EnvironmentSettings.newInstance().inStreamingMode().build());
+
+
+    // Start Kafka and Schema Registry Testcontainers. Set the exposed ports that test subclasses
+    // can use to dynamically configure Kafka connectors. Schema Registry enables connectors to
+    // be configured with 'value.format' = 'avro-confluent'
+    Network network = Network.newNetwork();
+
+    KafkaContainer kafka = new KafkaContainer(DockerImageName.parse("confluentinc/cp-kafka:7.3.2"))
+        .withEnv("KAFKA_TRANSACTION_STATE_LOG_REPLICATION_FACTOR", "1")
+        .withEnv("KAFKA_TRANSACTION_STATE_LOG_MIN_ISR", "1")
+        .withEnv("KAFKA_TRANSACTION_STATE_LOG_NUM_PARTITIONS", "1")
+        .withEnv("KAFKA_GROUP_INITIAL_REBALANCE_DELAY_MS", "500")
+        .withEnv("KAFKA_AUTO_CREATE_TOPICS_ENABLE", "true")
+        .withReuse(true)
+        .withNetwork(network);
+    kafka.start();
+    kafkaPort = kafka.getMappedPort(KAFKA_PORT);
+
+    GenericContainer schemaRegistry = new GenericContainer(DockerImageName.parse("confluentinc/cp-schema-registry:7.3.2"))
+        .withExposedPorts(8081)
+        .withNetwork(kafka.getNetwork())
+        .withEnv("SCHEMA_REGISTRY_HOST_NAME", "localhost")
+        .withEnv("SCHEMA_REGISTRY_LISTENERS", "http://0.0.0.0:8081")
+        .withEnv("SCHEMA_REGISTRY_KAFKASTORE_BOOTSTRAP_SERVERS", "PLAINTEXT://" + kafka.getNetworkAliases().get(0) + ":9092");
+    schemaRegistry.start();
+    schemaRegistryPort = schemaRegistry.getMappedPort(8081);
```

Comment:
It doesn't matter for this tutorial since there's only one test, but if we end up with more than one test we can put the container setup in a static block or a`@BeforeAll` so we do the container setup once for all tests - just a note for the future.